R/util.R

```diff
@@ -1635,6 +1635,72 @@ setdiff <- function(x, y, force_data_type = FALSE, ...){
   }
 }
 
+# This is written by removing unnecessary part from calculate_cohens_d.
+#'Calculate common standard deviation.
+#'@export
+calculate_common_sd <- function(var1, var2) {
+  df <- data.frame(var1=var1, var2=var2)
+  summarized <- df %>% dplyr::group_by(var2) %>%
+    dplyr::summarize(n=n(), v=var(var1, na.rm=TRUE))
+    
+  lx <- summarized$n[[1]] - 1
+  ly <- summarized$n[[2]] - 1
+  vx <- summarized$v[[1]]
+  vy <- summarized$v[[2]]
+  csd <- lx * vx + ly * vy
+  csd <- csd/(lx + ly)
+  csd <- sqrt(csd) # common sd computation
+}
+
+# Reference: https://stackoverflow.com/questions/15436702/estimate-cohens-d-for-effect-size
+#'Calculate Cohen's d
+#'@export
+calculate_cohens_d <- function(var1, var2) {
+  df <- data.frame(var1=var1, var2=var2)
+  summarized <- df %>% dplyr::group_by(var2) %>%
+    dplyr::summarize(n=n(), m=mean(var1, na.rm=TRUE), v=var(var1, na.rm=TRUE))
+    
+  lx <- summarized$n[[1]] - 1
+  ly <- summarized$n[[2]] - 1
+  mx <- summarized$m[[1]]
+  my <- summarized$m[[2]]
+  vx <- summarized$v[[1]]
+  vy <- summarized$v[[2]]
+  md  <- abs(mx - my) # mean difference (numerator)
+  csd <- lx * vx + ly * vy
+  csd <- csd/(lx + ly)
+  csd <- sqrt(csd) # common sd computation
+  cd  <- md/csd # cohen's d
+}
+
+# References:
+# SSb, SSt, eta squared definition: https://learningstatisticswithr.com/lsr-0.6.pdf
+# Cohen's f definition: https://en.wikipedia.org/wiki/Effect_size
+# Compared results with sjstats::cohens_f(), and powerAnalysis::ES.anova.oneway()
+# Did not use sjstats::cohens_f() to avoid requiring entire sjstats and its dependencies.
+# Did not use powerAnalysis::ES.anova.oneway() because it only works for the case all categories
+# have same number of observations.
+#'Calculate Cohen's f
+#'@export
+calculate_cohens_f <- function(var1, var2) {
+  m <- mean(var1, na.rm = TRUE)
+  df <- data.frame(var1=var1, var2=var2)
+  summarized <- df %>% dplyr::group_by(var2) %>%
+    dplyr::mutate(diff_between = mean(var1) - m, diff_total = var1 - m) %>% dplyr::ungroup() %>%
```

Comment:
Let's add na.rm=TRUE for mean and sum too.

Reply:
Done.